aw-server/src/endpoints/hostcheck.rs

```diff
@@ -0,0 +1,182 @@
+//! Host header check needs to be performed to protect against DNS poisoning
+//! attacks[1].
+//!
+//! Uses a Request Fairing to intercept the request before it's handled.
+//! If the Host header is not valid, the request will be rerouted to a
+//! BadRequest
+//!
+//! [1]: https://github.com/ActivityWatch/activitywatch/security/advisories/GHSA-v9fg-6g9j-h4x4
+use rocket::fairing::Fairing;
+use rocket::handler::Outcome;
+use rocket::http::uri::Origin;
+use rocket::http::{Method, Status};
+use rocket::{Data, Request, Rocket, Route};
+
+use crate::config::AWConfig;
+
+static FAIRING_ROUTE_BASE: &str = "/checkheader_fairing";
+
+pub struct HostCheck {
+    validate: bool,
+}
+
+impl HostCheck {
+    pub fn new(config: &AWConfig) -> HostCheck {
+        // We only validate requests if the server binds a local address
+        let validate = config.address == "127.0.0.1" || config.address == "localhost";
+        HostCheck { validate }
+    }
+}
+
+/// Route for HostCheck Fairing error
+fn fairing_error_route<'r>(_request: &'r Request<'_>, _: Data) -> Outcome<'r> {
+    Outcome::Failure(Status::BadRequest)
+}
```

Comment:
Any chance we could have a more informative error in this particular case of BadRequest? Just getting a BadRequest without any indication of why went wrong is kinda bad dev UX.

Reply:
Done